File: app/src/main/java/com/ella/music/ui/components/SquigglyProgress.java
package com.ella.music.ui.components;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.animation.ValueAnimator;
import android.content.res.ColorStateList;
import android.graphics.Canvas;
import android.graphics.ColorFilter;
import android.graphics.LinearGradient;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.PixelFormat;
import android.graphics.RadialGradient;
import android.graphics.RectF;
import android.graphics.Shader;
import android.graphics.drawable.Drawable;
import android.os.SystemClock;
import android.view.animation.PathInterpolator;

public final class SquigglyProgress extends Drawable {
    private static final float TWO_PI = 6.2831855f;

    private final Paint progressPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
    private final Paint linePaint = new Paint(Paint.ANTI_ALIAS_FLAG);
    private final Paint glowPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
    private final Paint headPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
    private final Path headPath = new Path();
    private final RectF glowRect = new RectF();

    private ValueAnimator heightAnimator;
    private boolean animate;
    private boolean transitionEnabled = true;
    private float heightFraction = 1f;
    private float lineAmplitude;
    private float phaseOffset;
    private float phaseSpeed;
    private float strokeWidth;
    private float waveLength;
    private long lastFrameTime = -1L;
    private int baseColor = 0xFFFFFFFF;
    private int baseAlpha = 255;

    public SquigglyProgress() {
        Paint.Cap cap = Paint.Cap.ROUND;
        progressPaint.setStrokeCap(cap);
        linePaint.setStrokeCap(cap);
        progressPaint.setStyle(Paint.Style.STROKE);
        linePaint.setStyle(Paint.Style.STROKE);
        glowPaint.setStyle(Paint.Style.FILL);
        headPaint.setStyle(Paint.Style.FILL);
        updateColors(baseColor, baseAlpha);
    }

    @Override
    public void draw(Canvas canvas) {
        if (animate) {
            invalidateSelf();
            long now = SystemClock.uptimeMillis();
            if (lastFrameTime > 0L) {
                phaseOffset += ((now - lastFrameTime) / 1000f) * Math.max(1f, phaseSpeed);
            }
            lastFrameTime = now;
        }

        float width = getBounds().width();
        float height = getBounds().height();
        if (width <= 0f || height <= 0f) return;

        float progress = clamp(getLevel() / 10000f, 0f, 1f);
        float centerY = getBounds().centerY();
        float left = getBounds().left;
        float right = left + width;
        float progressX = left + width * progress;
        float visible = clamp(heightFraction, 0f, 1f);
        float safeStroke = Math.max(1f, strokeWidth);
        float headLength = Math.max(safeStroke * 9f, Math.max(1f, waveLength) * 1.18f);
        float tailLength = Math.max(headLength * 4.6f, safeStroke * 24f);
        float glowHeight = Math.max(safeStroke * 5.5f, lineAmplitude * 2.2f) * visible;
        float pulse = animate ? (0.92f + 0.08f * (float) Math.sin((phaseOffset / 280f) * TWO_PI)) : 1f;

        linePaint.setStrokeWidth(safeStroke);
        progressPaint.setStrokeWidth(safeStroke);

        canvas.drawLine(left, centerY, right, centerY, linePaint);

        if (progressX > left) {
            progressPaint.setShader(new LinearGradient(
                left,
                centerY,
                progressX,
                centerY,
                new int[] {
                    setAlphaComponent(baseColor, (int) (baseAlpha * 0.62f * visible)),
                    setAlphaComponent(baseColor, (int) (baseAlpha * 0.92f * visible)),
                    setAlphaComponent(baseColor, (int) (baseAlpha * visible))
                },
                new float[] {0f, 0.72f, 1f},
                Shader.TileMode.CLAMP
            ));
            canvas.drawLine(left, centerY, progressX, centerY, progressPaint);
            progressPaint.setShader(null);
        }

        if (progress <= 0f) return;

        float tailStart = Math.max(left, progressX - tailLength);
        float glowEnd = Math.min(right, progressX + headLength * 0.72f);
        glowPaint.setShader(new LinearGradient(
            tailStart,
            centerY,
            glowEnd,
            centerY,
            new int[] {
                setAlphaComponent(baseColor, 0),
                setAlphaComponent(baseColor, (int) (baseAlpha * 0.22f * visible)),
                setAlphaComponent(baseColor, (int) (baseAlpha * 0.70f * visible * pulse)),
                setAlphaComponent(baseColor, 0)
            },
            new float[] {0f, 0.54f, 0.86f, 1f},
            Shader.TileMode.CLAMP
        ));
        glowRect.set(tailStart, centerY - glowHeight * 0.5f, glowEnd, centerY + glowHeight * 0.5f);
        canvas.drawRoundRect(glowRect, glowHeight * 0.5f, glowHeight * 0.5f, glowPaint);

        float headTip = Math.min(right, progressX + headLength * 0.46f);
        float headBack = Math.max(left, progressX - headLength * 0.74f);
        float headHalfHeight = Math.max(safeStroke * 2.2f, lineAmplitude * 0.62f) * visible * pulse;
        headPaint.setShader(new LinearGradient(
            headBack,
            centerY,
            headTip,
            centerY,
            new int[] {
                setAlphaComponent(baseColor, 0),
                setAlphaComponent(baseColor, (int) (baseAlpha * 0.82f * visible)),
                setAlphaComponent(baseColor, baseAlpha)
            },
            new float[] {0f, 0.58f, 1f},
            Shader.TileMode.CLAMP
        ));
        headPath.rewind();
        headPath.moveTo(headTip, centerY);
        headPath.cubicTo(progressX, centerY - headHalfHeight, headBack, centerY - headHalfHeight * 0.72f, headBack, centerY);
        headPath.cubicTo(headBack, centerY + headHalfHeight * 0.72f, progressX, centerY + headHalfHeight, headTip, centerY);
        headPath.close();
        canvas.drawPath(headPath, headPaint);

        glowPaint.setShader(new RadialGradient(
            progressX,
            centerY,
            Math.max(headLength * 1.25f, glowHeight),
            new int[] {
                setAlphaComponent(baseColor, (int) (baseAlpha * 0.45f * visible * pulse)),
                setAlphaComponent(baseColor, (int) (baseAlpha * 0.12f * visible)),
                setAlphaComponent(baseColor, 0)
            },
            new float[] {0f, 0.45f, 1f},
            Shader.TileMode.CLAMP
        ));
        glowRect.set(
            progressX - headLength * 1.4f,
            centerY - glowHeight,
            progressX + headLength * 1.1f,
            centerY + glowHeight
        );
        canvas.drawOval(glowRect, glowPaint);
        glowPaint.setShader(null);
        headPaint.setShader(null);
    }

    @Override
    public int getAlpha() {
        return baseAlpha;
    }

    public boolean getAnimate() {
        return animate;
    }

    public float getLineAmplitude() {
        return lineAmplitude;
    }

    @Override
    public int getOpacity() {
        return PixelFormat.TRANSLUCENT;
    }

    public float getPhaseSpeed() {
        return phaseSpeed;
    }

    public float getStrokeWidth() {
        return strokeWidth;
    }

    public boolean getTransitionEnabled() {
        return transitionEnabled;
    }

    public float getWaveLength() {
        return waveLength;
    }

    @Override
    protected boolean onLevelChange(int level) {
        return true;
    }

    @Override
    public void setAlpha(int alpha) {
        updateColors(baseColor, alpha);
    }

    public void setAnimate(boolean value) {
        if (animate == value) return;
        animate = value;
        if (value) {
            lastFrameTime = SystemClock.uptimeMillis();
        }

        if (heightAnimator != null) {
            heightAnimator.cancel();
        }
        ValueAnimator animator = ValueAnimator.ofFloat(heightFraction, animate ? 1f : 0f);
        if (animate) {
            animator.setStartDelay(60L);
            animator.setDuration(700L);
            animator.setInterpolator(new PathInterpolator(0.05f, 0.7f, 0.1f, 1f));
        } else {
            animator.setDuration(420L);
            animator.setInterpolator(new PathInterpolator(0f, 0f, 0.2f, 1f));
        }
        animator.addUpdateListener(valueAnimator -> {
            heightFraction = (Float) valueAnimator.getAnimatedValue();
            invalidateSelf();
        });
        animator.addListener(new AnimatorListenerAdapter() {
            @Override
            public void onAnimationEnd(Animator animation) {
                heightAnimator = null;
            }
        });
        animator.start();
        heightAnimator = animator;
    }

    @Override
    public void setColorFilter(ColorFilter colorFilter) {
        progressPaint.setColorFilter(colorFilter);
        linePaint.setColorFilter(colorFilter);
        glowPaint.setColorFilter(colorFilter);
        headPaint.setColorFilter(colorFilter);
    }

    public void setLineAmplitude(float value) {
        lineAmplitude = value;
    }

    public void setPhaseSpeed(float value) {
        phaseSpeed = value;
    }

    public void setStrokeWidth(float value) {
        if (strokeWidth == value) return;
        strokeWidth = value;
        progressPaint.setStrokeWidth(value);
        linePaint.setStrokeWidth(value);
    }

    @Override
    public void setTint(int color) {
        updateColors(color, getAlpha());
    }

    @Override
    public void setTintList(ColorStateList tint) {
        if (tint == null) return;
        updateColors(tint.getDefaultColor(), getAlpha());
    }

    public void setTransitionEnabled(boolean value) {
        transitionEnabled = value;
        invalidateSelf();
    }

    public void setWaveLength(float value) {
        waveLength = value;
    }

    public void updateColors(int color, int alpha) {
        baseColor = color;
        baseAlpha = clampInt(alpha, 0, 255);
        progressPaint.setColor(setAlphaComponent(baseColor, baseAlpha));
        linePaint.setColor(setAlphaComponent(baseColor, (int) ((baseAlpha / 255f) * 74)));
        glowPaint.setColor(setAlphaComponent(baseColor, baseAlpha));
        headPaint.setColor(setAlphaComponent(baseColor, baseAlpha));
        invalidateSelf();
    }

    private static int setAlphaComponent(int color, int alpha) {
        return ((clampInt(alpha, 0, 255) & 0xFF) << 24) | (color & 0x00FFFFFF);
    }

    private static float clamp(float value, float min, float max) {
        return Math.max(min, Math.min(max, value));
    }

    private static int clampInt(int value, int min, int max) {
        return Math.max(min, Math.min(max, value));
    }
}

File: app/src/main/java/com/ella/music/ui/components/SquigglyProgress.kt
package com.ella.music.ui.components

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.animation.ValueAnimator
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.ColorFilter
import android.graphics.Paint
import android.graphics.Path
import android.graphics.drawable.Drawable
import android.os.SystemClock
import android.view.animation.AccelerateDecelerateInterpolator
import android.view.animation.DecelerateInterpolator
import kotlin.math.abs
import kotlin.math.cos
import kotlin.math.max
import kotlin.math.min

/**
 * 波浪形进度条 Drawable（逆向自系统 UI SquigglyProgress）
 *
 * 绘制原理：
 * 1. 用贝塞尔曲线(cubicTo)绘制正弦波形路径
 * 2. 已播放部分：clipRect 裁剪 + wavePaint（不透明）绘制波浪
 * 3. 未播放部分：clipRect 裁剪 + linePaint（alpha=77/255 半透明）绘制波浪或直线
 * 4. 起点处绘制一个跳动的圆点（用余弦函数计算 Y 偏移）
 * 5. transitionEnabled=true 时，未播放部分也画波浪（彗星尾巴效果）
 * 6. animate=true 时，波浪相位随时间推移，产生流动动画
 */
class SquigglyProgress : Drawable() {

    var animate = false
        private set
    var heightFraction = 1f
    var lastFrameTime = -1L
    var lineAmplitude = 0f
    var phaseOffset = 0f
    var phaseSpeed = 0f
    var strokeWidth = 0f
        set(value) {
            if (field == value) return
            field = value
            wavePaint.strokeWidth = value
            linePaint.strokeWidth = value
        }
    var transitionEnabled = true
    var waveLength = 0f

    private val wavePaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        strokeCap = Paint.Cap.ROUND
        style = Paint.Style.STROKE
    }
    private val linePaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        strokeCap = Paint.Cap.ROUND
        style = Paint.Style.STROKE
        alpha = 77  // 半透明，产生渐隐效果
    }
    private val path = Path()
    private var heightAnimator: ValueAnimator? = null

    private val transitionPeriods = 1.5f
    private val minWaveEndpoint = 0.2f
    private val matchedWaveEndpoint = 0.6f

    override fun draw(canvas: Canvas) {
        drawTraced(canvas)
    }

    private fun drawTraced(canvas: Canvas) {
        // 相位动画：波浪随时间流动
        if (animate) {
            invalidateSelf()
            val now = SystemClock.uptimeMillis()
            phaseOffset = (((now - lastFrameTime) / 1000f) * phaseSpeed + phaseOffset) % waveLength
            lastFrameTime = now
        }

        // 计算播放进度（level 范围 0~1）
        var level = level / 10000f
        val width = bounds.width().toFloat()
        val progressX = width * level

        // 过渡区域映射：让低进度时波浪不展开到底
        if (transitionEnabled && level <= matchedWaveEndpoint) {
            level = lerp(minWaveEndpoint, matchedWaveEndpoint, lerpInv(0f, matchedWaveEndpoint, level))
        }
        val levelWidth = level * width

        // 振幅计算函数
        val computeAmplitude = { x: Float, sign: Float ->
            if (transitionEnabled) {
                val waveLen = waveLength * transitionPeriods
                val halfWave = waveLen / 2f
                val fadeFactor = lerpInvSat(levelWidth + halfWave, levelWidth - halfWave, x)
                lineAmplitude * sign * heightFraction * fadeFactor
            } else {
                lineAmplitude * sign * heightFraction
            }
        }

        // 用贝塞尔曲线绘制波浪路径
        val startOffset = -phaseOffset - waveLength / 2f
        val drawEnd = if (transitionEnabled) width else levelWidth
        val halfWave = waveLength / 2f

        path.rewind()
        path.moveTo(startOffset, 0f)
        var sign = 1f
        var prevAmplitude = computeAmplitude(startOffset, sign)
        var x = startOffset
        while (x < drawEnd) {
            sign = -sign
            val midX = x + halfWave
            val ctrlX = halfWave / 2f + x
            val nextAmplitude = computeAmplitude(midX, sign)
            path.cubicTo(ctrlX, prevAmplitude, ctrlX, nextAmplitude, midX, nextAmplitude)
            prevAmplitude = nextAmplitude
            x = midX
        }

        val clipHalf = lineAmplitude + strokeWidth

        canvas.save()
        canvas.translate(bounds.left.toFloat(), bounds.centerY().toFloat())

        // 已播放部分：实色波浪
        canvas.save()
        canvas.clipRect(0f, -clipHalf, progressX, clipHalf)
        canvas.drawPath(path, wavePaint)
        canvas.restore()

        // 未播放部分：半透明波浪（彗星尾巴）或直线
        if (transitionEnabled) {
            canvas.save()
            canvas.clipRect(progressX, -clipHalf, width, clipHalf)
            canvas.drawPath(path, linePaint)
            canvas.restore()
        } else {
            canvas.drawLine(progressX, 0f, width, 0f, linePaint)
        }

        // 起点圆点：随波浪上下跳动
        val pointY = cos(kotlin.math.abs(startOffset) / waveLength * 2 * Math.PI).toFloat() *
            lineAmplitude * heightFraction
        canvas.drawPoint(0f, pointY, wavePaint)

        canvas.restore()
    }

    override fun getAlpha(): Int = wavePaint.alpha

    override fun getOpacity(): Int = -3  // TRANSLUCENT

    override fun setAlpha(alpha: Int) {
        wavePaint.color = setAlpha(wavePaint.color, alpha)
        linePaint.color = setAlpha(wavePaint.color, ((alpha / 255f) * 77).toInt())
    }

    override fun setColorFilter(colorFilter: ColorFilter?) {
        wavePaint.colorFilter = colorFilter
        linePaint.colorFilter = colorFilter
    }

    override fun onLevelChange(level: Int): Boolean = animate

    fun setAnimate(value: Boolean) {
        if (animate == value) return
        animate = value
        if (value) lastFrameTime = SystemClock.uptimeMillis()

        heightAnimator?.cancel()
        val target = if (animate) 1f else 0f
        val animator = ValueAnimator.ofFloat(heightFraction, target).apply {
            if (animate) {
                startDelay = 60L
                duration = 800L
                interpolator = AccelerateDecelerateInterpolator()
            } else {
                duration = 550L
                interpolator = DecelerateInterpolator()
            }
            addUpdateListener {
                heightFraction = it.animatedValue as Float
                invalidateSelf()
            }
            addListener(object : AnimatorListenerAdapter() {
                override fun onAnimationEnd(animation: Animator) {
                    heightAnimator = null
                }
            })
            start()
        }
        heightAnimator = animator
    }

    override fun setTint(color: Int) {
        updateColors(color, alpha)
    }

    fun updateColors(color: Int, alpha: Int) {
        wavePaint.color = setAlpha(color, alpha)
        linePaint.color = setAlpha(color, ((alpha / 255f) * 77).toInt())
    }

    companion object {
        /** 设置颜色的 alpha 分量 */
        private fun setAlpha(color: Int, alpha: Int): Int =
            (alpha and 0xFF shl 24) or (color and 0x00FFFFFF)

        /** 线性插值 */
        private fun lerp(start: Float, end: Float, fraction: Float): Float =
            start + (end - start) * fraction.coerceIn(0f, 1f)

        /** 线性插值的逆函数 */
        private fun lerpInv(start: Float, end: Float, value: Float): Float {
            if (start == end) return 0f
            return ((value - start) / (end - start)).coerceIn(0f, 1f)
        }

        /** 饱和线性插值逆函数 */
        private fun lerpInvSat(start: Float, end: Float, value: Float): Float {
            val clamped = value.coerceIn(min(start, end), max(start, end))
            return lerpInv(start, end, clamped)
        }
    }
}
